### PR TITLE
#1981 - Request Offering Change: Ministry View/Approve/Deny (UI only) - Part 1

### PR DESCRIPTION
**Pull request by sh16011993**
## As a part of this PR, the following tasks were completed:

- [x] Created the models for Approve and Decline the application offering change request by the ministry.

Reason for modifying the base component (ModalDialogBase.vue): `<hr>` tag was changed to `<v-divider>` component. This v-divider component by default adds a significant amount of vertical whitespace. As a result, the margin for `v-divider` in this component is changed to reduce the whitespace.
 
Please note: Placeholder added for the API endpoint.

### Decline Application Offering Change Request Modal

image

### Approve Application Offering Change Request Modal

image

### CODE REVIEW

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 1]
```diff
@@ -0,0 +1,120 @@
+<template>
+  <v-form ref="assessApplicationOfferingChangeRequestForm">
+    <modal-dialog-base
+      :showDialog="showDialog"
+      :title="getTitle"
+      :max-width="780"
```
[the remark this answers is not in the document]
Comment (sh16011993) [reply sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:6]: The default `max-width` in the `modal-dialog-base` component is a bit less resulting in the wrapping of the line **"Outline the reasoning for approving this request. Please add the application number."** in the dialog. In the figma, the text is a single line. That is the reason I increased the `max-width`.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        <p class="label-bold my-2">Notes</p>
+        <v-textarea
+          variant="outlined"
+          hide-details="auto"
+          class="my-2"
+          v-model="note"
+          :rules="[checkNotesLengthRule]"
+          required
+        />
+        <p class="pt-1">
```
Comment (ann-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:32]: this is brand-grey, 
image

File: sources/packages/web/src/components/generic/ModalDialogBase.vue [hunk 3]
```diff
@@ -14,17 +14,17 @@
...
       </v-card-title>
-      <v-divider class="border-opacity-100" :thickness="2" inset />
-      <v-card-text class="pt-0">
+      <v-divider class="border-opacity-100 my-0" :thickness="2" inset />
+      <v-card-text class="pt-0 pb-1">
         <div class="pb-2" v-if="subTitle">{{ subTitle }}</div>
         <slot name="content">Please add the modal content here!</slot>
       </v-card-text>
-      <v-divider class="border-opacity-100" :thickness="2" inset />
+      <v-divider class="border-opacity-100 my-0" :thickness="2" inset />
```
Comment (ann-aot) [inline sources/packages/web/src/components/generic/ModalDialogBase.vue:27]: why this change in generic component?

Comment (sh16011993) [reply sources/packages/web/src/components/generic/ModalDialogBase.vue:27]: Without adding the `my-0` to the v-dividers, they are adding a default margin which increases a lot of vertical white space between the elements above and below it. `mt-3` was added to the title because there is no top margin otherwise and the component doesn't match Figma in that case.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    <modal-dialog-base
+      :showDialog="showDialog"
+      :title="getTitle"
+      :max-width="780"
+    >
+      <template #content
+        ><error-summary
+          :errors="assessApplicationOfferingChangeRequestForm.errors"
+        />
+        <p class="mt-4">
```
Comment (ann-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:12]: just want to double-check, do we really need these margins and padding with `p` tags? Lines 12, 23, 32

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue (hunk 2 above)
Comment (sh16011993) [reply sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue]: Updated it.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue (hunk 4 above)
Comment (sh16011993) [reply sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:12]: The problem is that the default vertical margin of the `v-divider` is too high. I changed to no margin in the common component and gave the margin as required in this component to the elements instead.

File: sources/packages/web/src/components/generic/ModalDialogBase.vue (hunk 3 above)
Comment (sh16011993) [reply sources/packages/web/src/components/generic/ModalDialogBase.vue]: I have the updated the `modal-dialog-base` component reducing the number of changes to the component and I have looked up many pre-existing modal dialogs. They look fine with the changes.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue (hunk 4 above)
Comment (sh16011993) [reply sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue]: Updated this component as well with the style changes.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.aest.controller.ts [hunk 5]
```diff
@@ -91,4 +100,19 @@ export class ApplicationOfferingChangeRequestAESTController extends BaseControll
       { hasAuditAndNoteDetails: true },
     );
   }
+
+  /**
+   * Updates an application offering change request.
+   * @param applicationOfferingChangeRequestId application offering change request id.
+   * @param payload information to update the application offering change request.
+   */
+  @Patch(":applicationOfferingChangeRequestId")
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.aest.controller.ts:109]: As discussed, missing the ministry role here. 

The todo comment description can mention the role as well.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.aest.controller.ts [hunk 6]
```diff
@@ -91,4 +100,19 @@ export class ApplicationOfferingChangeRequestAESTController extends BaseControll
       { hasAuditAndNoteDetails: true },
     );
   }
+
+  /**
+   * Updates an application offering change request.
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.aest.controller.ts:105]: The description can be more precise IMO(Approving or declining).

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/models/application-offering-change-request.dto.ts [hunk 7]
```diff
@@ -173,6 +174,14 @@ export class StudentApplicationOfferingChangeRequestAPIInDTO {
   applicationOfferingChangeRequestStatus: ApplicationOfferingChangeRequestStatus;
 }
 
+export class ApplicationOfferingChangeAssessmentAPIInDTO {
+  @IsNotEmpty()
+  @MaxLength(NOTE_MAX_LENGTH)
+  note: string;
+  @IsEnum(ApplicationOfferingChangeRequestStatus)
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/models/application-offering-change-request.dto.ts:181]: I missed to mention in the student part of the story. `@IsIn` will be more accurate here.

```ts
@IsIn([ApplicationOfferingChangeRequestStatus.Approved, ApplicationOfferingChangeRequestStatus.DeclinedBySABC]
```

File: sources/packages/backend/libs/sims-db/src/entities/application-offering-change-request.model.ts [hunk 8]
```diff
@@ -15,6 +15,7 @@ import { Note } from "./note.model";
 import { ApplicationOfferingChangeRequestStatus } from "./application-offering-change-request-status.type";
 
 export const REASON_MAX_LENGTH = 500;
+export const NOTE_MAX_LENGTH = 500;
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/sims-db/src/entities/application-offering-change-request.model.ts:18]: image

the notes max length is already available

File: sources/packages/web/src/components/generic/ModalDialogBase.vue [hunk 9]
```diff
@@ -14,17 +14,17 @@
...
         </slot>
       </v-card-title>
-      <v-divider class="border-opacity-100" :thickness="2" inset />
+      <v-divider class="border-opacity-100 my-1" :thickness="2" inset />
       <v-card-text class="pt-0">
         <div class="pb-2" v-if="subTitle">{{ subTitle }}</div>
         <slot name="content">Please add the modal content here!</slot>
       </v-card-text>
-      <v-divider class="border-opacity-100" :thickness="2" inset />
+      <v-divider class="border-opacity-100 my-1" :thickness="2" inset />
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/generic/ModalDialogBase.vue:27]: What is the reason for these changes in the base component? 

Also, if there is a valid reason for modifying the base component, please mention in the PR description.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrepestana-aot) [reply sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:6]: I think we need to keep all the modals with the same width unless we have any requirement. Otherwise the visual will be affected.

**Review by andrepestana-aot — COMMENTED** [review]
Good job. There are a few issues already pointed.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/models/application-offering-change-request.dto.ts [hunk 10]
```diff
@@ -173,6 +174,14 @@ export class StudentApplicationOfferingChangeRequestAPIInDTO {
   applicationOfferingChangeRequestStatus: ApplicationOfferingChangeRequestStatus;
 }
 
+export class ApplicationOfferingChangeAssessmentAPIInDTO {
+  @IsNotEmpty()
+  @MaxLength(NOTE_MAX_LENGTH)
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/models/application-offering-change-request.dto.ts:179]: If we are getting the max from the DB perspective it should come from the `NOTE_DESCRIPTION_MAX_LENGTH` that represents the DB max length for the table `sims.notes`.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+          <span
+            v-if="
+              showParameter ===
+              ApplicationOfferingChangeRequestStatus.DeclinedBySABC
+            "
+            >declining</span
+          ><span v-else>approving</span>
+          this request. Please add the application number.
+        </p>
+        <p class="label-bold mb-1">Notes</p>
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:23]: Figma is not targeting Vuetify. For the Vuetify component, we do not create extra labels we just use the out-of-box component `label` property even if it is a bit different from Figma.
https://github.com/bcgov/SIMS/blob/cf55ff4172c2dd87de6e44b6ee523084062a6e7f/sources/packages/web/src/components/aest/institution/modals/AssessOfferingModal.vue#L9-L15

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    <modal-dialog-base
+      :showDialog="showDialog"
+      :title="getTitle"
+      :max-width="780"
+    >
+      <template #content
+        ><error-summary
+          :errors="assessApplicationOfferingChangeRequestForm.errors"
+        />
+        <p class="my-4">
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:12]: I would prefer creating the entire text as a computed property and not individually replacing a word.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 13]
```diff
@@ -0,0 +1,119 @@
+<template>
+  <v-form ref="assessApplicationOfferingChangeRequestForm">
+    <modal-dialog-base
+      :showDialog="showDialog"
+      :title="getTitle"
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:5]: I see the `get` prefix used in the computed value, I would recommend leaving the `get` for methods.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+<script lang="ts">
+import ModalDialogBase from "@/components/generic/ModalDialogBase.vue";
+import { ref, defineComponent, computed } from "vue";
+import { useModalDialog, useRules } from "@/composables";
+import { ApplicationOfferingChangeAssessmentAPIInDTO } from "@/services/http/dto";
+import { ApplicationOfferingChangeRequestStatus, VForm } from "@/types";
+
+export default defineComponent({
+  components: {
+    ModalDialogBase,
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:55]: I believe that we talked in the past about having a common "Approval" modal dialog (similar to the [ConfimrModal](https://github.com/bcgov/SIMS/blob/6e3c85f97bcdd84284813b1fb1cc4c44201e8192/sources/packages/web/src/components/common/modals/ConfirmModal.vue)) that we allow adding a note and a decline/approve buttons but we never went for that. This situation would be perfect example to use a generic approval modal.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+  },
+  setup() {
+    const note = ref("");
+    const { showDialog, showModal, resolvePromise, showParameter } =
+      useModalDialog<ApplicationOfferingChangeAssessmentAPIInDTO | boolean>();
+    const assessApplicationOfferingChangeRequestModal = ref(
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:61]: Do we need this variable created as a `ref` or do we need this variable at all?
Can you please clarify why is it returned in the setup method?

File: sources/packages/web/src/views/aest/institution/ApplicationOfferingChangeRequestForm.vue [hunk 16]
```diff
@@ -147,6 +177,33 @@ export default defineComponent({
...
+      if (responseData) {
+        try {
+          await ApplicationOfferingChangeRequestService.shared.updateApplicationOfferingChangeRequestStatus(
+            props.applicationOfferingChangeRequestId,
+            responseData as ApplicationOfferingChangeAssessmentAPIInDTO,
+          );
+          router.push({
+            name: AESTRoutesConst.ASSESSMENTS_SUMMARY,
+          });
+          const snackbarMessage =
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/institution/ApplicationOfferingChangeRequestForm.vue:197]: Is the intention of having the `snackbarMessage` declared as a const to improve the code readability?
I am not sure is it is worth it in this scenario.

File: sources/packages/web/src/services/ApplicationOfferingChangeRequestService.ts [hunk 17]
```diff
@@ -190,7 +191,9 @@ export class ApplicationOfferingChangeRequestService {
    */
   async updateApplicationOfferingChangeRequestStatus(
     applicationOfferingChangeRequestId: number,
-    payload: StudentApplicationOfferingChangeRequestAPIInDTO,
+    payload:
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/ApplicationOfferingChangeRequestService.ts:194]: I would discourage using the same service method here. Both ministry and student operations have different DTO and there is no common component using this method.

**Review by andrewsignori-aot — COMMENTED** [review]
Noce start and thanks for breaking the PR. Please take a look at the comments.

File: sources/packages/web/src/services/http/ApplicationOfferingChangeRequestApi.ts [hunk 18]
```diff
@@ -192,7 +193,9 @@ export class ApplicationOfferingChangeRequestApi extends HttpBaseClient {
    */
   async updateApplicationOfferingChangeRequestStatus(
     applicationOfferingChangeRequestId: number,
-    payload: StudentApplicationOfferingChangeRequestAPIInDTO,
+    payload:
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/ApplicationOfferingChangeRequestApi.ts:196]: https://github.com/bcgov/SIMS/pull/2363/files#r1340722224

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+          required
+        />
+        <p class="pt-1 brand-gray-text">
+          Notes will be visible to StudentAid staff and institutions. This will
+          not be shown to students.
+        </p>
+      </template>
+      <template #footer>
+        <footer-buttons
+          :primaryLabel="getPrimaryLabel"
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:38]: https://github.com/bcgov/SIMS/pull/2363/files#r1340714293

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+      assessApplicationOfferingChangeRequestForm.value.resetValidation();
+      resolvePromise(false);
+    };
+    const assessChange = async () => {
+      const validationResult =
+        await assessApplicationOfferingChangeRequestForm.value.validate();
+      if (!validationResult.valid) {
+        return;
+      }
+      if (
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue:89]: What is the need for if else here?

this should work right?
`assessApplicationOfferingChangeRequestModal.value.applicationOfferingChangeRequestStatus = showParameter.value`

**Comment by dheepak-aot** [comment]
Nice start to the story. Please have a look at the comments.

**Review by ann-aot — COMMENTED** [review]
Please have a look at the comments.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.aest.controller.ts [hunk 21]
```diff
@@ -91,4 +100,19 @@ export class ApplicationOfferingChangeRequestAESTController extends BaseControll
...
+   * @param applicationOfferingChangeRequestId application offering change request id.
+   * @param payload information to update the application offering change request.
+   */
+  @Patch(":applicationOfferingChangeRequestId")
+  async updateApplicationOfferingChangeRequest(
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.aest.controller.ts:110]: Please add the the proper role for the method.

File: sources/packages/web/src/components/generic/ModalDialogBase.vue (hunk 9 above)
Comment (sh16011993) [reply sources/packages/web/src/components/generic/ModalDialogBase.vue:27]: Updated the reason for modifying the base component in the PR description.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.aest.controller.ts (hunk 21 above)
Comment (sh16011993) [reply sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/application-offering-change-request.aest.controller.ts:110]: Will be a part of the next PR. I have updated the TODO comment with the same.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue]: @sh16011993 I noticed the thumbs up from other DEVs here but I did not understand if the intention is to have `max-width="780"` removed or not.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes, just left one minor comment about one open conversation.

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (ann-aot) [reply sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue]: I don't see any issue with the default width. 
image
image

**Review by ann-aot — APPROVED** [review]
LGTM 👍 Just a minor comment . Thanks for doing the changes 👍 https://github.com/bcgov/SIMS/pull/2363#discussion_r1346597085

File: sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (sh16011993) [reply sources/packages/web/src/components/aest/AssessApplicationOfferingChangeRequestModal.vue]: Removed it.

File: sources/packages/web/src/components/generic/ModalDialogBase.vue (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/generic/ModalDialogBase.vue:27]: Thank you 👍

**Review by andrepestana-aot — APPROVED** [review]
Looks good. Thanks for the changes.

File: sources/packages/web/src/services/ApplicationOfferingChangeRequestService.ts [hunk 22]
```diff
@@ -197,4 +198,19 @@ export class ApplicationOfferingChangeRequestService {
       payload,
     );
   }
+
+  /**
+   * Approves or declines the application offering change request status.
+   * @param applicationOfferingChangeRequestId application offering change request id.
+   * @param payload information to update the application offering change request.
+   */
+  async updateAESTApplicationOfferingChangeRequestStatus(
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/ApplicationOfferingChangeRequestService.ts:207]: The name `updateAESTApplicationOfferingChangeRequestStatus` doesn't seem to be appropriate. when ministry approves we do more than updating status. I would suggest `AssessApplicationOfferingChangeRequest`

File: sources/packages/web/src/services/http/ApplicationOfferingChangeRequestApi.ts [hunk 23]
```diff
@@ -197,4 +198,17 @@ export class ApplicationOfferingChangeRequestApi extends HttpBaseClient {
     const url = `application-offering-change-request/${applicationOfferingChangeRequestId}`;
     await this.patchCall(this.addClientRoot(url), payload);
   }
+
+  /**
+   * Approve or decline the application offering change request status.
+   * @param applicationOfferingChangeRequestId application offering change request id.
+   * @param payload information to update the application offering change request status and save the declaration.
+   */
+  async updateAESTApplicationOfferingChangeRequestStatus(
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/ApplicationOfferingChangeRequestApi.ts:207]: https://github.com/bcgov/SIMS/pull/2363/files#r1346620330

File: sources/packages/web/src/views/aest/institution/ApplicationOfferingChangeRequestForm.vue [hunk 24]
```diff
@@ -109,6 +134,11 @@ export default defineComponent({
     const studentApplicationOfferingDetails = ref(
       {} as ApplicationOfferingDetails,
     );
+    const assessApplicationOfferingChangeRequestModal = ref(
+      {} as ModalDialog<ApplicationOfferingChangeAssessmentAPIInDTO>,
```
Comment (dheepak-aot) [inline sources/packages/web/src/views/aest/institution/ApplicationOfferingChangeRequestForm.vue:138]: should it not be `ModalDialog<ApplicationOfferingChangeAssessmentAPIInDTO | false>`

as we return `resolvePromise(false)` on cancel ?

image

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes. 👍
